R/modified_update.R

```diff
@@ -532,3 +537,98 @@ mu_calculate_rate <- function(name, dat, window, n_years) {
 
   dat
 }
+
+mu_fix_sdf7_psu <- function(meta, data){
+  ### this function fixes SDF7 - PSU - Measles - Campaign impacts
+  i <- meta$group$touchstone_src == "201210gavi-201303gavi-1" & 
+    meta$group$modelling_group == "PSU-Ferrari" &
+    meta$group$vaccine == "Measles"
+  dat_id <- meta$group$index[i]
+  stopifnot(length(dat_id) == 2)
+  
+  ## keep the rest
+  keep <- data[!(data$index %in% dat_id), ]
+  ## identify problem source
+  dat1 <- data[data$index == dat_id[1], ]
+  dat2 <- data[data$index == dat_id[2], ]
+  ## operation
+  dat <- dat1
+  i <- is.na(dat1$deaths_averted) | dat1$deaths_averted == 0.
+  
+  dat$deaths_averted[i] <- dat2$deaths_averted[i]
+  dat$cases_averted[i] <- dat2$cases_averted[i]
+  dat$fvps[i] <- dat2$fvps[i]
+  dat$coverage_old[i] <- dat2$coverage_old[i]
+  dat$coverage_target[i] <- dat2$coverage_target[i]
+  
+  
+  i <- sum(dat$deaths_averted, na.rm = TRUE) - 
+    (sum(dat1$deaths_averted, na.rm = TRUE) + sum(dat2$deaths_averted, na.rm = TRUE))
+  stopifnot( abs(i) < 1 )
+  
+  ### re-calculate impact rates
+  i <- is_blank(dat$fvps) &
+    !(is_blank(dat$deaths_averted) & is_blank(dat$cases_averted)) &
+    !is_blank(dat$coverage_target)
+  if (any(i)) {
+    stopifnot(x$activity_type == "campaign")
+    message("Creating synthetic fvps for ", x$model)
+    ## Otherwise use pop_routine.
+    dat$fvps[i] <- dat$coverage_old[i] * dat$coverage_target[i]
+  }
+  
+  i <- is_blank(dat$fvps) & !is_blank(dat$coverage_old)
+  if (any(i)) {
+    message("Creating synthetic fvps for ", x$model)
+    if((x$activity_type == "campaign")){
+      dat$fvps[i] <- dat$coverage_old[i] * dat$coverage_target[i]
+    }else{
+      dat$fvps[i] <- dat$coverage_old[i] * dat$pop_routine[i]
+    }
+  }
+  
+  year_min <- 2001
+  year_max <- 2030
+  w <- 4L # 1/2 window for rolling average
+  window <- 2 * w + 1
+  years <- year_min:year_max
+  n_years <- length(years)
+  
+  dat <- mu_calculate_rate("deaths", dat, window, n_years)
+  dat <- mu_calculate_rate("cases", dat, window, n_years)
+  
+  dat$target_pop_estimated <- dat$fvps / dat$coverage_old
+  dat$target_pop_estimated[dat$fvps > 0 & is_blank(dat$coverage_old)] <- NA
+  
+  dat$target_pop_estimated_avg <-
+    roll_mean_by(dat$target_pop_estimated, window, n_years)
+  
+  
+  dat$target_pop_given <- dat$coverage_target
+  dat$target_pop_given_new <- dat$coverage_target_new
+  
+  
+  ## We also want new fvps which we compute as coverage * target pop
+  v <- "coverage_target_new"
+  dat$fvps_new <- dat$coverage_new * dat[[v]]
+  
+  ## At the moment we can do nothing for MHL, TUV and XK, because we have no routine pop for them
+  i <- !is_blank(dat$coverage_new) & is_blank(dat$fvps_new) & dat$year < 2031#& !is_blank(dat$target_pop_estimated)
```

Comment:
Can the commented out target_pop_estimated come out?

Reply:
In fact this bit of code is irrelevant to campaigns. I have deleted them.

Reply:
How much did you delete?  Will this break doing modified updates for previous cases?

Reply:
I deleted this line and the immediate if(...) clause. Previously the issue only applies to routine.

Reply:
Either I don't understand what you're doing or you are not understanding my problem with this change.  

* The code that you have removed was needed for a previous modified update to run. 
* Presumably it still is.
* So if you delete it, then in order to allow computing a modified update on this one particular touchstone, the package can no longer compute modified updates on previous touchstones
* We will have to compute modified updates against old touchstones in the future
* So when we do that next time you'll add this line back in or we'll have to constantly install particular versions.

This is not going to be a reasonable way to continue

Reply:
This bit of code lies within the "mu_fix_sdf7_psu()" function. It would only be executed when we update SDF7 with another touchstone. It would not affect anything else. 

I have put he code back, so that it is more consistent with what was done.